staging/src/k8s.io/apiserver/pkg/storage/names/generate_test.go

```diff
@@ -19,11 +19,50 @@ package names
 import (
 	"strings"
 	"testing"
+
+	utilrand "k8s.io/apimachinery/pkg/util/rand"
 )
 
 func TestSimpleNameGenerator(t *testing.T) {
-	name := SimpleNameGenerator.GenerateName("foo")
-	if !strings.HasPrefix(name, "foo") || name == "foo" {
-		t.Errorf("unexpected name: %s", name)
+	mkBase := func(n int) string {
+		return utilrand.String(n-1) + "-"
+	}
+
+	cases := []struct {
+		base        string
+		expectLen   int
+		expectTrunc int
+	}{{
+		// Normal use
+		base:      mkBase(3),
+		expectLen: 3 + randomLength,
+	}, {
+		// Edge case: long name, just enough for suffix
+		base:      mkBase(maxNameLength - randomLength),
+		expectLen: maxNameLength,
+	}, {
+		// Edge case: long name trucated
+		base:        mkBase(maxNameLength - (randomLength - 1)),
+		expectLen:   maxNameLength,
+		expectTrunc: 1,
+	}, {
+		// Edge case: max length name trucated
+		base:        mkBase(maxNameLength),
+		expectLen:   maxNameLength,
+		expectTrunc: randomLength,
+	}, {
+		// Edge case: very long name trucated
+		base:        mkBase(maxNameLength * 2),
+		expectLen:   maxNameLength,
+		expectTrunc: maxNameLength + randomLength,
+	}}
+	for i, tc := range cases {
+		name := SimpleNameGenerator.GenerateName(tc.base)
+		if len(name) != tc.expectLen {
+			t.Errorf("case[%d]: wrong result len: expected %d, got %d (%q)", i, tc.expectLen, len(name), name)
+		}
+		if pfx := tc.base[0 : len(tc.base)-tc.expectTrunc]; !strings.HasPrefix(name, pfx) {
```

Comment:
This verifies that it didn't truncate more than it should, perhaps for proper paranoia we should verify that it also didn't truncate less than it should?

Reply:
I can check that the result is not the same as the input, but anything is hard to ensure, unless the end of the input is known to be characters not in the random set.  I could set the first truncated base character to something not in the random charset.

Will add a commit.